### DISCUSSION TITLE
Merging old (SQLite) databases

### DISCUSSION THREAD

**Discussion by DD4WH** [mcguirepr89/BirdNET-Pi#294] [Q&A]
I would like to merge my old databases that have been produced by old versions of BirdNET-Pi.

1.) MariaDB databases
* as far as I know, these databases are the files called `BirdDB.txt`
* what do I have to do to import these into my DBBrowser SQLite ?
* what do I have to do to export these into the desired current SQLite format ?

2.) old SQLite databases (as far as I remember, from February 2022)
* I tried to just swap this database `birds.db` into a current installation, but they are not read by the current version

![grafik](https://user-images.githubusercontent.com/14326464/169324020-8d5e991c-4405-4cd5-af58-9e36061b33d9.png)

* I tried an old database from end of March and this is read successfully
* When I compared both database structures, they seem exactly the same

This is the February database (not read)
![grafik](https://user-images.githubusercontent.com/14326464/169324349-7a9325cb-abff-4553-882f-e7457f053acc.png)

This is the March database (read)
![grafik](https://user-images.githubusercontent.com/14326464/169324466-a0d977d9-78bd-4a56-9866-492177595144.png)

* What am I missing here? Is there a difference in those databases, that can only be seen in the DBBrowser SQLite when I switch on a specific option?

**Comment by mcguirepr89** [comment]
Hi, @DD4WH -- I'll need to look closely at this to determine what I will recommend to merge these data sets.

If I were to take a guess at what the difference is between those two databases, I would suspect they differ in _ownership_ and/or _permissions_.

The best way to tell who owns a file and what its permissions are (surprise surprise) is through the terminal. If you put both databases into the same directory and then issue `ls -l` in that directory, you'll get output similar to my example below:
```
birdnetpi1:~/BirdNET-Pi/scripts$ ls -l
total 14672
-rw-rw-r-- 1 birder birder   194401 Apr 12 14:18 adminer-de.php
-rw-rw-r-- 1 birder birder   195144 Apr 12 14:18 adminer-fr.php
-rw-rw-r-- 1 birder birder   193022 Apr 12 14:18 adminer.php
-rw-r--r-- 1 birder birder    14019 May 19 09:51 advanced.php
-rwxr-xr-x 1 birder birder     3680 May 19 09:52 analyze.py
-rwxr-xr-x 1 birder birder     5230 May 18 17:39 birdnet_analysis.sh
-rwxrwxr-x 1 birder birder    10463 Apr 24 10:28 birdnet_analysis.sh.tempfixes
-rwxr-xr-x 1 birder birder      200 May 19 09:51 birdnet_log.sh
-rwxrwxr-x 1 birder birder      154 May 12 16:59 birdnet_log.sh~
-rw-rw-r-- 1 birder birder      357 Apr 12 14:18 birdnetpi_request.php
-rwxrwxr-x 1 birder birder      943 May  5 16:57 birdnet_recording.sh
-rw-rw-r-- 1 birder birder  3588096 May 19 11:25 birds.db
-rw-rw-r-- 1 birder birder      356 Apr 12 14:18 birdweather_request.php
-rwxrwxr-x 1 birder birder      293 Apr 12 14:18 cleanup.sh
-rwxrwxr-x 1 birder birder     2593 Apr 12 14:18 clear_all_data.sh
-rw-r--r-- 1 birder birder    12419 May 19 09:51 config.php
. . . 
```

You'll notice that the `birds.db` database is owned by `birder` and belongs to the group` birder`. For its permissions, it can be read/written to by the _owner_ and the _group_ and can be read by everyone. This is a helpful [cheatsheet](https://infohubblog.com/linux-file-permissions-cheat-sheet.html) that you can use when reviewing your output. 

How it _should be_:
The `birds.db` file should be owned by the default $USER and belong to that user's GROUP. Likewise, the database should have _group_ read/write permissions.

To ensure these things, this example command will enforce that ownership and those permissions (assuming the database that needs changes to its ownership and permissions is located at `~/BirdNET-Pi/scripts/birds.db`):
```
sudo -E chown $USER:$USER ~/BirdNET-Pi/scripts/birds.db
sudo -E chmod g+rw ~/BirdNET-Pi/scripts/birds.db
```

@CaiusX's suggestion that everything makes its way into the same CSV and then import the CSV back into a new database seems like a reasonable (though tasking) method to use. I will think more on other means of merging this data that may be slightly less tedious.

One word of advice: BACK EVERYTHING UP!!!! Keep a copy of each database somewhere in case attempts to merge the data lead to corruption.

**Reply by DD4WH** [reply]
Thanks a lot, Patrick!!! Sorry for the late reaction.
I will have to try your suggestions and will then report back here!

**Comment by oliverk71** [comment]
I am also looking for an quick, easy and reliable way to merge SQLite databases. 
Uploading a file (copy of birds.db) via file manager did not work for me as well, although it is the same owner (oliver in my case). It only has additional executive rights, but that shouldn't be the cause. 
`-rwxrwxrwx 1 oliver  2,8M Mai 22 17:25 birds(1)_.db`

**Reply by mcguirepr89** [reply]
I think I know the cause behind this. 
If you would like to try restoring your database, could you please try again using these steps to confirm if I am correct in my suspicions? 

1. "Tools" > "Services" > "Disable" Pushed Notications
2. "Tools" > "File Manager" > Delete the current `$HOME/BirdNET-Pi/scripts/birds.db` file
3. Again using the "File Manager", Upload the backup `birds.db` you wish to restore into the same `$HOME/BirdNET-Pi/scripts` directory, effectively replacing the `birds.db` removed in step 2.

If that works, the Pushed Notification service is the "bug" in file manager (obviously not a bug in Tiny File Manager, but in BirdNET-Pi -- don't worry, the Pushed Notification service is being removed in lieu of the new Apprise notification services).

**Comment by oliverk71** [comment]
Unfortunately it does not work properly. But I should mention that I just renamed the `birds.db` to `birds_fewdaystomerge.db` and did not delete it. However, I there is a file `birds.db`, now, but it has a size of 0 B.

**Reply by mcguirepr89** [reply]
Yes that is being created by the `pushed_notification.service`, which is why I asked that you disable it before trying to replace the database. I'm curious how the empty database was created with the service disabled???????/

You can try this instead:
```
stop_core_services.sh
sudo systemctl disable --now pushed_notifications.service
sudo rm /lib/systemd/system/pushed_notifications.service
sudo rm ~/BirdNET-Pi/scripts/birds.db
```

Now there should NOT be a `birds.db` file in `$HOME/BirdNET-Pi/scripts`

**Reply by oliverk71** [reply]
I am pretty sure I disabled it. But I can try what you suggested in your last post. 
Weird... `sudo rm ~/BirdNET-Pi/scripts/birds.db` did not delete the file, but `rm birds.db` did. Edit: No, it did not. 

![grafik](https://user-images.githubusercontent.com/10213957/169905331-e34b145f-a880-4b05-8154-46db8e5dfb3b.png)

**Reply by mcguirepr89** [reply]
I'm sure both did, but that it was recreated instantly. Issuing `ls -l` will be helpful when using it to determine those things, as the timestamp will let you know.

**Reply by oliverk71** [reply]
```
birdnetpi-elsen:~/BirdNET-Pi/scripts$ ls -gh | grep .db
-rw-r--r-- 1 caddy     0 May 23 23:02 birds.db
-rw-rw-r-- 1 oliver  60K May 23 21:08 birds_fewdaystomerge.db
-rwxrwxr-x 1 oliver  510 May 22 20:06 createdb.sh
birdnetpi-elsen:~/BirdNET-Pi/scripts$ sudo rm birds.db
birdnetpi-elsen:~/BirdNET-Pi/scripts$ ls -gh | grep .db
-rw-r--r-- 1 caddy     0 May 23 23:03 birds.db
-rw-rw-r-- 1 oliver  60K May 23 21:08 birds_fewdaystomerge.db
-rwxrwxr-x 1 oliver  510 May 22 20:06 createdb.sh
birdnetpi-elsen:~/BirdNET-Pi/scripts$ 
```

**Reply by mcguirepr89** [reply]
Yes, both commands are removing the file, which is being recreated instantaneously by the `pushed_notification.service`. It needs to be disabled _before_ allll of this.

```
sudo systemctl disable --now pushed_notifications.service
sudo systemctl stop pushed_notifications.service
sudo rm /lib/systemd/system/pushed_notifications.service
sudo rm ~/BirdNET-Pi/templates/pushed_notifications.service
sudo rm ~/BirdNET-Pi/scripts/birds.db
```

Please issue ALL of the commands above in that order without skipping any command.

**Reply by oliverk71** [reply]
Well ...
```
birdnetpi-elsen:~/BirdNET-Pi/scripts$ sudo systemctl disable --now pushed_notifications.service
sudo systemctl stop pushed_notifications.service
sudo rm /lib/systemd/system/pushed_notifications.service
sudo rm ~/BirdNET-Pi/templates/pushed_notifications.service
sudo rm ~/BirdNET-Pi/scripts/birds.db
Failed to disable unit: Unit file pushed_notifications.service does not exist.
Failed to stop pushed_notifications.service: Unit pushed_notifications.service not loaded.
rm: cannot remove '/lib/systemd/system/pushed_notifications.service': No such file or directory
birdnetpi-elsen:~/BirdNET-Pi/scripts$ 
```

**Comment by oliverk71** [comment]
Well ...
```
birdnetpi-elsen:~/BirdNET-Pi/scripts$ sudo systemctl disable --now pushed_notifications.service
sudo systemctl stop pushed_notifications.service
sudo rm /lib/systemd/system/pushed_notifications.service
sudo rm ~/BirdNET-Pi/templates/pushed_notifications.service
sudo rm ~/BirdNET-Pi/scripts/birds.db
Failed to disable unit: Unit file pushed_notifications.service does not exist.
Failed to stop pushed_notifications.service: Unit pushed_notifications.service not loaded.
rm: cannot remove '/lib/systemd/system/pushed_notifications.service': No such file or directory
birdnetpi-elsen:~/BirdNET-Pi/scripts$ 
```

**Reply by mcguirepr89** [reply]
Good, now did any `birds.db` file magically appear?

**Reply by oliverk71** [reply]
```
birdnetpi-elsen:~/BirdNET-Pi/scripts$ ls -gh | grep .db
-rw-r--r-- 1 caddy     0 May 23 23:07 birds.db
-rw-rw-r-- 1 oliver  60K May 23 21:08 birds_fewdaystomerge.db
-rwxrwxr-x 1 oliver  510 May 22 20:06 createdb.sh

```

**Reply by mcguirepr89** [reply]
Thank you, now we know that the pushed_notification.service is not the root cause and can continue troubleshooting.

To confirm, this needs to be tried while the core services are stopped, since if they are running, that could be creating the empty database in an attempt to write to it.

Please try the following:
On your Linux Mint host machine with the backup of `birds.db` in your `Downloads` directory:
```
scp ~/Downloads/birds.db oliver@birdnetpi-elsen.local:/home/oliver/birds.db
```

and then on your BirdNET-Pi:
```
mv -f ~/birds.db ~/BirdNET-Pi/scripts/birds.db
```

I am very confident this will work, as it is the [instruction set from the wiki](https://github.com/mcguirepr89/BirdNET-Pi/wiki/Backup-and-Restore-the-Database).

Please let me know if that does not work for you.

**Reply by oliverk71** [reply]
That worked!

```
birdnetpi-elsen:~/BirdNET-Pi/scripts$ mv -f ~/birds.db ~/BirdNET-Pi/scripts/birds.db
birdnetpi-elsen:~/BirdNET-Pi/scripts$ ls -l | grep db
-rwxr-xr-x 1 oliver oliver  2924544 May 23 23:24 birds.db
-rw-rw-r-- 1 oliver oliver    61440 May 23 21:08 birds_fewdaystomerge.db
-rwxrwxr-x 1 oliver oliver      510 May 22 20:06 createdb.sh
```

**Reply by oliverk71** [reply]
Regarding merge: Would this work in the SQLite:
```
[attach] '~/BirdNET-Pi/scripts/birdstomerge.db' as toMerge;           
[BEGIN]; 
[insert] into AuditRecords [select](https://www.sqlite.org/lang_select.html) * from toMerge.AuditRecords; 
[COMMIT]; 
[detach] toMerge;
```
I did not yet try.

**Reply by mcguirepr89** [reply]
Hi, Oliver -- I have no idea about the commands above. I'm ashamed to admit I don't even know what programming language that is :/ sorry

**Reply by oliverk71** [reply]
Hi Patrick, 
in "Database Maintenance" there is a link "SQL commands", where commands can be entered and executed. 
See for example here: https://www.sqlitetutorial.net/sqlite-attach-database/

**Comment by DD4WH** [comment]
After some months I made another attempt to merge two SQLite databases. 

I have two databases from my BirdNET-Pi:

Database1.db contains my detections from 25.2.2022 to 12.6.2022
Database2.db contains my detections from 12.6.2022 to 9.08.2022

Here is my workflow to merge these two database and put the merged database back to my BirdNET-Pi:

1.) Download SQLite Studio and open the program
2.) Add Database1.db, add Database2.db
3.) In Database2.db, click on table detections and then on "data"
4.) Click the "Export" button which will open a dialogue: export data, leave all other settings as they are, choose a sensible name for the csv-file
5.) click on table detections in Database1.db
6.) click on then icon "Import data", a dialogue opens
7.) leave all settings as they are and choose the file you have exported a few seconds ago. Import takes a long time
8.) Now your databases have been merged and the Database1.db now contains all the detections from both databases!
9.) Go to your BirdNET-Pi --> Tools --> Servies --> Stop Core Services
10.) Open WinSCP (if not yet installed, install it NOW) and log into your BirdNET-Pi
11.) navigate to BirdNET-Pi/scripts and rename birds.db to "birds_old.db"
12.) Copy Birds1.db (the merged database) from your local computer to your BirdNET-Pi/scripts folder
13.) Rename Birds1.db into "birds.db"
14.) Go to your BirdNET-Pi --> Web Terminal
15.) Issue 
`sudo -E chown $USER:$USER ~/BirdNET-Pi/scripts/birds.db`
`sudo -E chmod g+rw ~/BirdNET-Pi/scripts/birds.db`
16.) Done!
17.) Test if it has worked by going to the "Overview" page and look at the number of detections. This should now include ALL detections from both databases.

Please write feedback, I have surely made some errors in my description . . . its not a very easy task, I can tell you ;-)

I have tested this successfully with two VERY large databases containing a little less than half a million detections after the merge ;-).

![grafik](https://user-images.githubusercontent.com/14326464/183976337-0a844313-ada0-430f-92f2-bcc00eaa9e90.png)

Example of Blackcap detections from End of February 2022 up to now (the two humps in the phenology diagram can be interpreted as the occurence of two generations in this season)
![grafik](https://user-images.githubusercontent.com/14326464/183974001-093ab8ea-1a47-417a-b0df-3b4c34d9f12a.png)

Thanks to Patrick for his help with ownership and assigning write rights for the database file !!!
